powerauth-push-server/src/main/java/io/getlime/push/controller/rest/PushDeviceController.java

```diff
@@ -93,17 +95,32 @@ public Response createDevice(@RequestBody ObjectRequest<CreateDeviceRequest> req
         }
         logger.info("Received createDevice request, app ID: {}, activation ID: {}, token: {}, platform: {}", requestObject.getAppId(),
                 requestObject.getActivationId(), maskPushToken(requestObject.getToken()), requestObject.getPlatform());
-        String errorMessage = CreateDeviceRequestValidator.validate(requestObject);
+        final String errorMessage = CreateDeviceRequestValidator.validate(requestObject);
         if (errorMessage != null) {
             throw new PushServerException(errorMessage);
         }
-        String appId = requestObject.getAppId();
-        String pushToken = requestObject.getToken();
-        String platform = requestObject.getPlatform();
-        String activationId = requestObject.getActivationId();
-        List<PushDeviceRegistrationEntity> devices = lookupDeviceRegistrations(appId, activationId, pushToken);
-        final AppCredentialsEntity appCredentials = findAppCredentials(appId);
-        PushDeviceRegistrationEntity device;
+
+        final AppCredentialsEntity appCredentials = findAppCredentials(requestObject.getAppId());
+
+        // In case of parallel requests to create new device, the createOrUpdateDevice may fail because of unique
+        // constraint violation. Try to repeat the save.
+        RetryTemplate.builder()
```

Comment:
Have you considered using`@Retryable` and moving it into the service layer instead of the controller?

Reply:
Will be considered in the separate issue https://github.com/wultra/powerauth-push-server/issues/742